lib/custom_failure.rb

```diff
@@ -13,7 +13,11 @@ def redirect
   end
 
   def redirect_url
-    root_path
+    if request.get?
+      attempted_path
+    else
+      request.referrer
+    end
```

Comment:
Aaron was working on this before in order to deal with some SSO issues:
https://github.com/concord-consortium/rigse/pull/52
I think the main things to test are:
1. what happens when your portal session is expired and you try to log in through the portal from LARA or some other OAuth client.
2. what happens when your portal session is expired and you try to go somewhere that requires authentication.
3. what happens when your portal session is expired and you try to save a form of some kind (POST request). Also check what happens when this form is submitted via XHR.
4. recheck #2 and #3 when the cookie has been completely deleted (which should trigger a warden message other than `:timeout`)